### PR TITLE
BF16 autocast (speed up epochs, fit more training in 5 min)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Our current model runs at ~31s/epoch, fitting only 10 epochs in the 5-minute budget. BF16 mixed precision (torch.amp.autocast with bfloat16) can significantly speed up the forward and backward passes — especially the attention computations and linear layers. If we can reduce epoch time to ~20-25s, we get 12-15 epochs instead of 10, which means more gradient updates and better convergence.

This is the simplest possible speed optimization — no architecture changes, just wrapping the compute in autocast. BF16 is well-suited because it has the same exponent range as FP32 (no overflow issues) while halving memory bandwidth.

## Instructions

In `train.py`, wrap the forward pass and loss computation in the training loop with autocast (around lines 141-149):

**Before:**
```python
        pred = model({"x": x, "is_surface": is_surface})["preds"]
        sq_err = (pred - y_norm) ** 2
        vol_mask = mask & ~is_surface
        surf_mask = mask & is_surface
        vol_loss = (sq_err * vol_mask.unsqueeze(-1)).sum() / vol_mask.sum().clamp(min=1)
        channel_w = torch.tensor([1.0, 1.0, 1.5], device=device)
        surf_loss = surface_loss_curriculum(pred, y_norm, surf_mask, channel_w, epoch, MAX_EPOCHS)
        loss = vol_loss + cfg.surf_weight * surf_loss
```

**After:**
```python
        with torch.amp.autocast("cuda", dtype=torch.bfloat16):
            pred = model({"x": x, "is_surface": is_surface})["preds"]
            sq_err = (pred - y_norm) ** 2
            vol_mask = mask & ~is_surface
            surf_mask = mask & is_surface
            vol_loss = (sq_err * vol_mask.unsqueeze(-1)).sum() / vol_mask.sum().clamp(min=1)
            channel_w = torch.tensor([1.0, 1.0, 1.5], device=device)
            surf_loss = surface_loss_curriculum(pred, y_norm, surf_mask, channel_w, epoch, MAX_EPOCHS)
            loss = vol_loss + cfg.surf_weight * surf_loss
```

Do the same for the validation loop (around lines 184-191):
```python
        with torch.amp.autocast("cuda", dtype=torch.bfloat16):
            pred = model({"x": x, "is_surface": is_surface})["preds"]
            sq_err = (pred - y_norm) ** 2
```

That's it — everything else stays the same. Same model, same hyperparams, same loss. Just faster.

W&B tag: `mar14b`, group: `bf16-autocast`

## Baseline
- **surf_p = 107.35**, surf_Ux = 1.23, surf_Uy = 0.84, val_loss = 2.45
- ~31s/epoch, 10 epochs in 5 min

---

## Results
_To be filled by student_

### CODE REVIEW

**Comment by tcapelle** [comment]
## Results

**W&B run:** bvyqj4r5

| Metric | Baseline | This run |
|--------|----------|----------|
| surf_p | 107.35 | **174.8** |
| surf_Ux | 1.23 | 1.83 |
| surf_Uy | 0.84 | 1.05 |
| val/loss | 2.45 | 2.68 |

**Speed improvement confirmed:** 23s/epoch vs ~31s baseline (26% faster), VRAM 12.2GB vs 15.4GB (21% less). 13 epochs in 5.1 min vs ~10 before.

**Outcome: Negative** — despite more epochs, surf_p degraded by ~67 points vs baseline.

Training was highly unstable under BF16 (validation spikes at epochs 4, 6, 10-11). The reduced numerical precision of BF16 appears to destabilize the physics attention computation, causing validation oscillations that prevent consistent convergence. The extra 3 epochs from the speedup did not compensate for the quality degradation.
